Accumulate notifications instead of blinking

PVince81 opened the pull request:
This makes it possible to display multiple notifications.
If the options.type is set to "error", it will also add a close button.

To test, edit "apps/files/ajax/upload.php" and add these lines to simulate failure:

```
    OCP\JSON::error(array('data' => ['message' => 'Connection timeout or other error message']));
    exit();
```

Then upload several files.

Another way is to call the API directly in the browser console:

```
OC.Notification.showHtml('<div>Oh hello</div><div>Second line</div>', {type: 'error'});
OC.Notification.show('test permanent notif without button');
OC.Notification.showTemporary('test temporary notif without button');
```

Some issues to discuss:
- [x] is the scrollbar acceptable when there are too many notifications ?
- [x] should upload errors time out ? (they used to, but it's convenient to be able to read it and close them afterwards)
- [x] missing "close all", how and where ?
- [x] should we always show the close button, even for non-errors ? (and force API consumers to pass "type: permanent" if they really want to make it unclosable ?)

Todos:
- [x] TODO: make sure the file name is included in all file-related error messages
- [x] TODO: write unit tests

Tests:
- [x] TEST: manual test (see example calls above)
- [x] TEST: upload with not enough space
- [x] TEST: users "undo" notification still works
- [x] TEST: update notification still works

@jancborchardt please try it out and see the items to discuss above

@MorrisJobke @rullzer @nickvergessen @schiesbn @icewind1991 

Fixes #19817

Commit:
Accumulate notifications instead of blinking

This makes it possible to display multiple notifications.
If the options.type is set to "error", it will also add a close button.

PVince81 commented:
Reposting the screenshot:
![cumulative-notifications](https://cloud.githubusercontent.com/assets/277525/13174030/62a9878e-d701-11e5-99c7-f678a4718203.png)

jancborchardt commented:
Looks fine :+1:

PVince81 commented:
Thanks.

Considering that this situation with lots of message should seldom happen (only in error cases), then it might be fine to NOT have a "close all" button. In the worst case the user can refresh the page.

nickvergessen commented:
JS tests still failing

PVince81 commented:
I haven't adjusted them yet. Wanted to make sure that the approach was ok.

I'll continue working on this then.

jancborchardt commented:
Yeah, this is an edge-case anyway, close all is not really needed.

nickvergessen commented:
Looks good, but shows the scrollbar in firefox although not needed:
![dateien_-_owncloud_-_2016-02-22_08 41 04](https://cloud.githubusercontent.com/assets/213943/13212267/12c89110-d940-11e5-8710-3c58b9815310.png)

Commit:
Fix unit tests affected by side effects

The notification tests were not restoring the clock properly, but
indirectly helped other tests pass.

Since now we're restoring the clock properly, the other tests were fixed
to still work.

Commit:
Remove unused update-notification style

PVince81 commented:
Unit tests written.

@nickvergessen fixed the scrollbar in Firefox (needed to be "auto", not "scroll")

I got rid of the "update-notification" CSS class because the new update notification code actually uses `OC.Notification.showTemporary` instead of having its own comfortable container.

PVince81 commented:
Fixes https://github.com/owncloud/core/issues/19817

nickvergessen commented:
Seems to work,
one last issue I noticed, when you drop multiple folders into the file list, only one `Error uploading file "Sync": Unable to upload Sync as it is a directory or has 0 bytes` message shows up.
I assume it's because of the endpoint and not JS, but would be nice to fix it at some point.

:+1:

MorrisJobke commented:
Tested and works :+1:
